controller/tests/controller_server/csi_controller_server_test.py

```diff
@@ -532,6 +545,62 @@ def test_create_volume_with_zero_size(self, storage_agent):
         self.assertEqual(self.context.code, grpc.StatusCode.OK)
         self.mediator.create_volume.assert_called_once_with(self.request.name, 1 * 1024 * 1024 * 1024, {}, "pool1")
 
+    def _prepare_idempotent_tests(self):
+        self.mediator.get_volume = Mock()
+        self.mediator.copy_to_existing_volume_from_source = Mock()
+        self.request.volume_content_source = self._get_source_snapshot("wwn1")
+
+    @patch("controller.controller_server.csi_controller_server.get_agent")
+    def test_create_volume_from_source_idempotent(self, storage_agent):
+        self._prepare_idempotent_tests()
+        storage_agent.return_value = self.storage_agent
+        snapshot_id = "wwn1"
+        self.mediator.get_volume.return_value = utils.get_mock_mediator_response_volume(10, volume_name, "wwn2", "a9k",
+                                                                                        copy_source_id=snapshot_id)
+
+        response = self.servicer.CreateVolume(self.request, self.context)
+
+        self.assertEqual(self.context.code, grpc.StatusCode.OK)
+        self.assertEqual(response.volume.content_source.snapshot.snapshot_id, snapshot_id)
+        self.mediator.copy_to_existing_volume_from_source.assert_not_called()
+
+    @patch("controller.controller_server.csi_controller_server.get_agent")
+    def test_create_volume_from_source_idempotent_no_source(self, storage_agent):
+        self._prepare_idempotent_tests()
+        storage_agent.return_value = self.storage_agent
+        self.mediator.get_volume.return_value = utils.get_mock_mediator_response_volume(10, volume_name, "wwn2",
+                                                                                        "a9k")
+        response = self.servicer.CreateVolume(self.request, self.context)
+
+        self.assertEqual(self.context.code, grpc.StatusCode.ALREADY_EXISTS)
+        self.assertEqual(response.volume.content_source.snapshot.snapshot_id, '')
+        self.mediator.copy_to_existing_volume_from_source.assert_not_called()
+
+    @patch("controller.controller_server.csi_controller_server.get_agent")
+    def test_create_volume_from_source_idempotent_source_not_requested_but_found(self, storage_agent):
+        self._prepare_idempotent_tests()
+        storage_agent.return_value = self.storage_agent
+        snapshot_id = "wwn1"
+        self.request.volume_content_source = None
+        self.mediator.get_volume.return_value = utils.get_mock_mediator_response_volume(10, volume_name, "wwn2", "a9k",
+                                                                                        copy_source_id=snapshot_id)
+        response = self.servicer.CreateVolume(self.request, self.context)
+
+        self.assertEqual(self.context.code, grpc.StatusCode.ALREADY_EXISTS)
+        self.assertEqual(response.volume.content_source.snapshot.snapshot_id, '')

```

Comment:
shouldn't there be an id value in this case?
and if not, should the response even contain a volume?
and if not, then why is the assert so deep in the response object?

Reply:
[spec ](https://github.com/container-storage-interface/spec/blob/v1.3.0/spec.md#createvolume)specify new created so "old" volume wont return

Reply:
done (changed to `HasField`)

Reply:
so why not [here](https://github.com/IBM/ibm-block-csi-driver/blob/0ccea31ebcc4c1b34b97674c0d065b8027ab2a17/controller/tests/controller_server/csi_controller_server_test.py#L576) also?

Reply:
right done

Reply:
and [here](https://github.com/IBM/ibm-block-csi-driver/blob/0b40468bbb83f62464aa6a483ad080ad590d91a7/controller/tests/controller_server/csi_controller_server_test.py#L403)?

Reply:
yep

Reply:
I don't see the change in my last link